File: archive/cpp/2021/nov/sum-root-to-leaf-numbers.cpp
#include <numeric>
#include <vector>

#include "common.h"

using namespace std;

class Solution
{
public:
    int sumNumbers(TreeNode *root)
    {
        vector<int> v;
        int cur = 0;
        helper(root, cur, &v);
        return accumulate(v.begin(), v.end(), 0);
    }

private:
    void helper(TreeNode *node, int cur, vector<int> *v)
    {
        if (!node)
        {
            return;
        }
        cur = cur * 10 + node->val;
        if (!node->left && !node->right)
        {
            v->push_back(cur);
        }
        if (node->left)
        {
            helper(node->left, cur, v);
        }
        if (node->right)
        {
            helper(node->right, cur, v);
        }
    }
};

File: src/tree/sum_root_to_leaf_numbers.rs
use crate::{Solution, TreeNode};
use std::cell::RefCell;
use std::rc::Rc;

impl Solution {
    pub fn sum_numbers(root: Option<Rc<RefCell<TreeNode>>>) -> i32 {
        let mut r = 0;
        fn iter_sum(root: &Option<Rc<RefCell<TreeNode>>>, n: &mut i32) -> i32 {
            match root {
                None => {
                    return 0;
                }
                Some(node) => {
                    *n = 10 * *n + node.borrow().val;

                    if node.borrow().left == node.borrow().right {
                        return *n;
                    }
                    return iter_sum(&node.borrow().left, n) + iter_sum(&node.borrow().right, n);
                }
            }
        }
        iter_sum(&root, &mut r);
        r
    }
}

#[cfg(test)]
mod tests {
    use super::Solution;

    #[test]
    fn test() {
        println!("{}", Solution::sum_numbers(None));
    }
}
